File: macos/pixel_compressor/Sources/pixel_compressor/ThumbnailEngine.swift
import AVFoundation
import CoreGraphics
import CoreMedia
import Foundation

/// Implements `ThumbnailHostApi.generateThumbnails`: extracts an exact
/// frame per requested position with `AVAssetImageGenerator`, resizes via
/// the shared `ImageFormatCodec` helper (reused from the image pipeline),
/// and encodes each frame independently.
final class ThumbnailEngine {
  private let cacheManager: CacheManager
  private let taskRegistry: TaskRegistry
  private let progressReporter: ProgressReporter

  init(cacheManager: CacheManager, taskRegistry: TaskRegistry, progressReporter: ProgressReporter) {
    self.cacheManager = cacheManager
    self.taskRegistry = taskRegistry
    self.progressReporter = progressReporter
  }

  func generate(request: ThumbnailRequest) throws -> [ThumbnailResultMessage] {
    let taskId = request.taskId

    try checkCancelled(taskId)
    progressReporter.reportProgress(taskId: taskId, stage: .preparing, percent: 0)

    guard FileManager.default.fileExists(atPath: request.sourcePath) else {
      throw PixelCompressorError.invalidMedia("Source file does not exist: \(request.sourcePath)")
    }
    let sourceURL = URL(fileURLWithPath: request.sourcePath)
    let asset = AVURLAsset(url: sourceURL)

    guard !asset.tracks(withMediaType: .video).isEmpty else {
      throw PixelCompressorError.unsupportedMedia("No video track found in source")
    }

    let generator = AVAssetImageGenerator(asset: asset)
    generator.appliesPreferredTrackTransform = true
    generator.requestedTimeToleranceBefore = .zero
    generator.requestedTimeToleranceAfter = .zero

    let format = request.format
    _ = try ImageFormatCodec.utType(for: format)  // reject WebP up front

    let maxEdge = max(Int(request.maxWidth), Int(request.maxHeight))
    let totalItems = request.positionsMs.count
    var results: [ThumbnailResultMessage] = []
    results.reserveCapacity(totalItems)

    for (index, positionMs) in request.positionsMs.enumerated() {
      try checkCancelled(taskId)

      let time = CMTime(value: positionMs, timescale: 1000)
      let cgImage: CGImage
      do {
        cgImage = try generator.copyCGImage(at: time, actualTime: nil)
      } catch {
        throw PixelCompressorError.decodingError(
          "Could not extract frame at \(positionMs)ms: \(error.localizedDescription)")
      }

      let resized = try ImageFormatCodec.resize(cgImage, maxPixelSize: maxEdge)

      let fileName = "\(taskId)_\(index).\(ImageFormatCodec.fileExtension(for: format))"
      let outputPath = try cacheManager.resolveOutputPath(kind: .thumbnails, fileName: fileName)
      let outputURL = URL(fileURLWithPath: outputPath)
      try? FileManager.default.removeItem(at: outputURL)

      try ImageFormatCodec.encode(
        image: resized, to: outputURL, format: format, quality: Int(request.quality))

      results.append(
        ThumbnailResultMessage(
          outputPath: outputURL.path,
          positionMs: positionMs,
          width: Int64(resized.width),
          height: Int64(resized.height)
        ))

      let percent = Double(index + 1) / Double(max(totalItems, 1)) * 100
      progressReporter.reportProgress(
        taskId: taskId, stage: .encoding, percent: percent,
        currentItemIndex: Int64(index + 1), totalItems: Int64(totalItems))
    }

    progressReporter.reportProgress(taskId: taskId, stage: .completed, percent: 100, force: true)
    return results
  }

  private func checkCancelled(_ taskId: String) throws {
    if taskRegistry.isCancelled(taskId: taskId) {
      throw PixelCompressorError.cancelled()
    }
  }
}

File: android/src/main/kotlin/com/therivanta/pixelcompressor/thumbnail/ThumbnailEngine.kt
package com.therivanta.pixelcompressor.thumbnail

import android.content.Context
import android.media.MediaMetadataRetriever
import com.therivanta.pixelcompressor.CompressionStageWire
import com.therivanta.pixelcompressor.ThumbnailRequest
import com.therivanta.pixelcompressor.ThumbnailResultMessage
import com.therivanta.pixelcompressor.cache.CacheKind
import com.therivanta.pixelcompressor.cache.CacheManager
import com.therivanta.pixelcompressor.errors.PixelCompressorErrors
import com.therivanta.pixelcompressor.image.BitmapTransform
import com.therivanta.pixelcompressor.image.ImageFormatCodec
import com.therivanta.pixelcompressor.progress.ProgressReporter
import java.io.File
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

/** Extracts frames from a video at requested timestamps via
 * [MediaMetadataRetriever.getFrameAtTime], reusing [BitmapTransform] for
 * resizing and [ImageFormatCodec] for encoding.
 */
class ThumbnailEngine(
  private val context: Context,
  private val cacheManager: CacheManager,
  private val progressReporter: ProgressReporter,
) {

  suspend fun generate(request: ThumbnailRequest): List<ThumbnailResultMessage> {
    val taskId = request.taskId
    val sourceFile = File(request.sourcePath)
    if (!sourceFile.exists()) {
      throw PixelCompressorErrors.invalidMedia("Source file does not exist: ${request.sourcePath}")
    }
    if (!sourceFile.canRead()) {
      throw PixelCompressorErrors.permissionDenied("Cannot read source file: ${request.sourcePath}")
    }

    progressReporter.onProgress(taskId, CompressionStageWire.PREPARING, 0.0)

    val retriever = MediaMetadataRetriever()
    try {
      try {
        retriever.setDataSource(request.sourcePath)
      } catch (e: Exception) {
        throw PixelCompressorErrors.unsupportedMedia(
          "Source is not a decodable video: ${request.sourcePath}",
          e.toString(),
        )
      }

      val scratchDir = File(context.cacheDir, "pixel_compressor/tmp").apply { mkdirs() }
      val total = request.positionsMs.size
      val results = mutableListOf<ThumbnailResultMessage>()
      val maxWidth = request.maxWidth.toInt()
      val maxHeight = request.maxHeight.toInt()
      val quality = request.quality.toInt().coerceIn(1, 100)

      for ((index, positionMs) in request.positionsMs.withIndex()) {
        currentCoroutineContext().ensureActive()

        val frame = try {
          retriever.getFrameAtTime(positionMs * 1000, MediaMetadataRetriever.OPTION_CLOSEST)
        } catch (e: Exception) {
          throw PixelCompressorErrors.decodingError(
            "Failed to extract frame at ${positionMs}ms: ${e.message}",
            e.toString(),
          )
        } ?: throw PixelCompressorErrors.decodingError("No frame available at ${positionMs}ms")

        val fitted = BitmapTransform.fitToBox(frame, maxWidth, maxHeight)
        val width = fitted.width
        val height = fitted.height
        val bytes = try {
          ImageFormatCodec.encode(fitted, request.format, quality, scratchDir)
        } finally {
          fitted.recycle()
        }

        val outputPath = cacheManager.resolveOutputPath(
          CacheKind.THUMBNAILS,
          ImageFormatCodec.extensionFor(request.format),
        )
        File(outputPath).writeBytes(bytes)

        results.add(
          ThumbnailResultMessage(
            outputPath = outputPath,
            positionMs = positionMs,
            width = width.toLong(),
            height = height.toLong(),
          ),
        )

        val itemNumber = (index + 1)
        progressReporter.onProgress(
          taskId,
          CompressionStageWire.ENCODING,
          itemNumber.toDouble() / total * 100.0,
          currentItemIndex = itemNumber.toLong(),
          totalItems = total.toLong(),
        )
      }

      progressReporter.onProgress(taskId, CompressionStageWire.COMPLETED, 100.0)
      return results
    } finally {
      retriever.release()
    }
  }
}
